feat: Support `GROUP BY` with tags in raw `SELECT` queries

stuartcarnie opened the pull request:
Closes #6903

Teach the planner how to process a query that specifies a `GROUP BY` clause with one or more tags.

# Background

This PR brings some clarity to the following issues.

## Sort order

Ensure the ordering of the result set is congruent with the existing InfluxQL implementation, which varies when specifying multiple measurements or when tags are specified in the column projection or `GROUP BY` clause. This ordering is established by the following

https://github.com/influxdata/influxdb_iox/blob/d36f8e04631a18ed7679ec8b370c023ac28a922c/iox_query/src/plan/influxql/planner.rs#L306-L314

## Group key identification

Provide metadata to a Flight client such that it is capable of decoding the results into various InfluxQL formats, such as JSON or columnar. 

Group key information is encoded in schema-level metadata using a structure defined in the `Flight.proto` IDL:

https://github.com/influxdata/influxdb_iox/blob/d36f8e04631a18ed7679ec8b370c023ac28a922c/generated_types/protos/influxdata/iox/querier/v1/flight.proto#L58-L61

This information allows a client to determine the index of the measurement column:

https://github.com/influxdata/influxdb_iox/blob/d36f8e04631a18ed7679ec8b370c023ac28a922c/generated_types/protos/influxdata/iox/querier/v1/flight.proto#L62

and the indices of the tag keys:

https://github.com/influxdata/influxdb_iox/blob/d36f8e04631a18ed7679ec8b370c023ac28a922c/generated_types/protos/influxdata/iox/querier/v1/flight.proto#L69-L72

in the Arrow `RecordBatch` instances.

### Propagating metadata

> **Note**
> 
> There is currently an issue that schema-level metadata is dropped when encoding to Flight due to the following issue:
>
> - https://github.com/apache/arrow-rs/issues/3779

Additionally, as discussed with @alamb, physical planning provides no guarantee that metadata is propagated, so a new physical operator was created to ensure this metadata was propagated:

https://github.com/influxdata/influxdb_iox/blob/d36f8e04631a18ed7679ec8b370c023ac28a922c/iox_query/src/frontend/influxql.rs#L54-L60

This operator wraps the physical plan returned by DataFusion and merges the schema-level metadata from the logical plan:

https://github.com/influxdata/influxdb_iox/blob/d36f8e04631a18ed7679ec8b370c023ac28a922c/iox_query/src/frontend/influxql.rs#L131-L144

The schema is then directly fetched via this call in the grpc module:

https://github.com/influxdata/influxdb_iox/blob/d36f8e04631a18ed7679ec8b370c023ac28a922c/service_grpc_flight/src/lib.rs#L687

Similar to the Arrow issue listed previously, a small change to the grpc module was applied per commit 3fd0f34, to ensure the schema-level metadata isn't lost.

> **Note**
>
> Running a modified version of Arrow with the fix described in https://github.com/apache/arrow-rs/issues/3779#issuecomment-1451046234 validated that metadata passes over flight.

wolffcm reviewed (approved):
You requested a review from Andrew but I thought that I would have a look too. I had some questions, but otherwise this looks right to me.

I have some thoughts about how this will integrate with gap filling, I will leave them on #6112.

stuartcarnie commented on generated_types/protos/influxdata/iox/querier/v1/flight.proto:
```diff
@@ -54,3 +54,20 @@ message ReadInfo {
+  // schema. The vector is sorted by the `name` field in lexicographically
+  // ascending order.
+  repeated TagKeyColumnIndex tag_key_columns = 2;
+}
```
This is still required, as we need to encode the Group Key information in the Arrow schema metadata to inform an InfluxQL client. Storing it in the Arrow schema means that it always accompanies the `RecordBatch`, regardless of how it travels across processes. I didn't want to store it as unstructured data in the Arrow schema metadata, so I formalised it as a Protobuf message, and serialise it as JSON. I was contemplating adding a format specifier as another key in the Arrow schema metadata, so we could specify the encoding format (JSON, protobuf, etc)

stuartcarnie commented on iox_query/src/plan/influxql/planner.rs:
```diff
@@ -1266,6 +1618,105 @@ mod test {
+                Projection: Dictionary(Int32, Utf8("disk")) AS iox::measurement, disk.time, CAST(disk.device AS Utf8) AS device, CAST(NULL AS Utf8) AS cpu, CAST(NULL AS Float64) AS usage_idle, NULL AS free [iox::measurement:Dictionary(Int32, Utf8), time:Timestamp(Nanosecond, None), device:Utf8;N, cpu:Utf8;N, usage_idle:Float64;N, free:Null;N]
+                  TableScan: disk [bytes_free:Int64;N, bytes_used:Int64;N, device:Dictionary(Int32, Utf8);N, host:Dictionary(Int32, Utf8);N, region:Dictionary(Int32, Utf8);N, time:Timestamp(Nanosecond, None)]
+            "###);
+        }
```
Yes, a `GROUP BY` with no aggregates essentially allows you to control how the data is sorted, but more importantly, affects how the data is rendered in InfluxQL's "pretty" format.

stuartcarnie commented on influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql:
```diff
@@ -258,4 +258,22 @@ SELECT f64, non_existing, f64 + non_existing FROM m0 WHERE f64 > 19;
+SELECT usage_idle, free FROM cpu, disk GROUP BY cpu;
+SELECT usage_idle, free FROM cpu, disk GROUP BY cpu, device;
+SELECT usage_idle, free FROM cpu, disk GROUP BY device, cpu;
+SELECT usage_idle, free, device, cpu FROM cpu, disk GROUP BY device, cpu;
```
Good catch, fixed them! 🙇🏻

stuartcarnie commented on iox_query/src/plan/influxql/planner.rs:
```diff
@@ -840,6 +1113,75 @@ mod test {
+                    sorted.sort_unstable();
+                    assert_eq!(keys, sorted, "tag keys are not sorted");
+                };
+            }
```
Another reason they are handy is the backtrace is at the call site of the macro

stuartcarnie commented:
Thanks again for the review, @wolffcm! I've fixed the test, good spot!

Please note that I added f6c07f86 to add an additional `is_projected` field to the InfluxQL metadata, with additional tests.

alamb commented on generated_types/protos/influxdata/iox/querier/v1/flight.proto:
```diff
@@ -54,3 +54,28 @@ message ReadInfo {
 // information in the future.
 message AppMetadata {}
+
+// A structure which describes the layout of the group key in a `RecordBatch`.
```
```suggestion
// A structure which describes the layout of the group key in a `RecordBatch`.
// This information is used to map the data in a `RecordBatch` to the InfluxDB data model
// where in addition to a data type, each columns is either a `tag`, `field` or `timestamp`
```

alamb commented on influxdb_influxql_parser/src/expression/arithmetic.rs:
```diff
@@ -874,19 +875,19 @@ mod test {
 
         // No arguments
-        assert_call("FN()", "FN()");
+        assert_call("FN()", "fn()");
```
I assume that the automatic lowercasing follows the InfluxQL rules? 

If so, it is also consistent with DataFusion's SQL rules: https://github.com/apache/arrow-datafusion/blob/c56b947651d923d99f9756f09af9c013430fc9c0/datafusion/sql/src/utils.rs#L541-L546

I believe one common confusion from people using SQL for the first time in IOx is that sql identifers are automatically lowercased.

So if you have a measurement named `myMeasurement`

The SQL 
```sql
select * from myMeasurement
```

Will tell you that `mymeasurement` doesn't exist. You have to put the identifier in double quotes

```sql
select * from "myMeasurement"
```

I wonder if this is the same for InfluxQL?

stuartcarnie replied:
This automatic lowercasing was to normalise the function calls, as they are case-insensitive identifiers. This allows us to safely perform case-sensitive lookups in a hash table, rather than repeatedly having to lower-case the function names to find matches. This is the same behaviour in InfluxQL's parser.

InfluxQL does not do automatic lowercasing of other identifiers, so in your example, the following would match exactly the measurement `myMeasurement` in InfluxQL:

```sql
SELECT * FROM myMeasurement
```

Quoting in InfluxQL is strictly for allowing spaces and other characters

I have to admit I don't recall ever hitting case-sensitivity in SQL, and I've been writing it since circa 1991 🤯  Perhaps I always used snake-case for my tables 😆

alamb commented on influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected:
```diff
@@ -409,8 +409,117 @@
+| disk             | 2022-10-31T02:00:10Z |            | 319833.0   |
+| disk             | 2022-10-31T02:00:10Z |            | 419833.0   |
++------------------+----------------------+------------+------------+
+-- InfluxQL: SELECT cpu, usage_idle FROM cpu;
```
Something that might be worth doing for these end to end tests, if the metadata is important to render properly in the client, would be to print out the metadata in the output as well 

Something like

```
-- InfluxQL: SELECT cpu, usage_idle FROM cpu;
Metadata: tag_columns: [2]
+------------------+----------------------+-----------+------------+
| iox::measurement | time                 | cpu       | usage_idle |
+------------------+----------------------+-----------+------------+
| cpu              | 2022-10-31T02:00:00Z | cpu-total | 2.98       |
| cpu              | 2022-10-31T02:00:00Z | cpu0      | 0.98       |
| cpu              | 2022-10-31T02:00:00Z | cpu1      | 1.98       |
| cpu              | 2022-10-31T02:00:10Z | cpu-total | 2.99       |
| cpu              | 2022-10-31T02:00:10Z | cpu0      | 0.99       |
| cpu              | 2022-10-31T02:00:10Z | cpu1      | 1.99       |
+------------------+----------------------+-----------+------------+
```

stuartcarnie replied:
That is a good idea – I'll do that 👍🏻

stuartcarnie replied:
@alamb I realise I will have to write a follow up issue and PR to do this, as we will need https://github.com/apache/arrow-rs/issues/3779

stuartcarnie replied:
Requisite issue:

- https://github.com/influxdata/influxdb_iox/issues/7145

alamb commented on iox_query/Cargo.toml:
```diff
@@ -25,6 +25,7 @@ datafusion = { workspace = true }
 datafusion_util = { path = "../datafusion_util" }
 executor = { path = "../executor"}
 futures = "0.3"
+generated_types = { path = "../generated_types" }
```
`iox_query` is already pretty large and `generated_types` brings in a bunch more dependencies.

I wonder if it is time to take a pause now and pull out the InfluxQL into a `iox_influxql` crate? I can probably find time to help with that next week

stuartcarnie replied:
Would you like me to pull it out in a separate PR or in this one?

stuartcarnie replied:
Relevant issue:

- https://github.com/influxdata/influxdb_iox/issues/7153

alamb commented on influxdb_iox/tests/query_tests2/cases/in/issue_6112.influxql.expected:
```diff
@@ -409,8 +409,117 @@
+| disk             | 2022-10-31T02:00:00Z |           | disk1s5 |            | 3234.0     |
+| disk             | 2022-10-31T02:00:10Z |           | disk1s5 |            | 3239.0     |
++------------------+----------------------+-----------+---------+------------+------------+
+-- InfluxQL: SELECT usage_idle, bytes_free FROM cpu, disk GROUP BY device, cpu;
```
I didn't realize you could select from several measurements like this (and it doesn't do a cross join like in SQL but instead appends the results). Nice

wolffcm replied:
I've heard that joins in InfluxQL are an often-requested feature. If we ever were to support them, I guess we'd have to come up with a new syntax.

alamb replied:
If we did I would recommend following the newer (SQL-99?) style JOIN syntax rather than the implicit cross join semantics.
```sql
SELECT  *
FROM 
  foo JOIN bar ON (foo.x = bar.y)
```

It is less potentially confusing as well as supports OUTER joins, etc

alamb commented on iox_query/src/frontend/influxql.rs:
```diff
@@ -43,6 +51,61 @@ impl SchemaProvider for ContextSchemaProvider {
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        unimplemented!()
```
Likewise here I think it would be good to avoid the panic 
Something like (untested):

```suggestion
assert!(children.len(), 1);
Ok(Arc::new(Self {
  schema: Arc::clone(&self.schema),
  input: children.pop(),
}))
```

alamb commented on iox_query/src/plan/influxql/planner.rs:
```diff
@@ -34,13 +38,20 @@ use itertools::Itertools;
 use std::str::FromStr;
 use std::sync::Arc;
 
+/// The name of the column specifying the source measurement for a row.
```
What would you think about moving these value (as well as the code that added them to/extracted them from the schema)  into the `schema` crate along with the other metadata (e.g.

https://github.com/influxdata/influxdb_iox/blob/8771dcb645055bce75a6d1f6169694e75e86fd71/schema/src/lib.rs#L154-L156

The rationale is that then the schema handling in a single location (and it would keep the influxql implementation less complicated)

stuartcarnie replied:
Yes, I like that – as it definitely feels misplaced here. I'll do that 👍🏻

stuartcarnie replied:
Done

stuartcarnie replied:
I left the `MEASUREMENT_COLUMN_INDEX` here as that is specified in the `InfluxQlMetadata` structure that the client must read.

alamb reviewed (approved):
Looks good to me! Thanks for the review as well @wolffcm.

stuartcarnie commented on iox_query/src/plan/influxql/planner.rs:
```diff
@@ -813,7 +1077,21 @@ mod test {
+            .schema()
+            .metadata()
+            .get(METADATA_KEY)
+            .map(|s| serde_json::from_str(s).unwrap())
```
Given this is only used in the tests, and we know it is valid, given we validate that it serialises correctly in the planner, it should be fine.

We do handle errors when deserialising the metadata, such as in the CLI.

stuartcarnie commented on iox_query/src/plan/influxql/planner.rs:
```diff
@@ -597,14 +727,82 @@ impl<'a> InfluxQLToLogicalPlan<'a> {
+///
+/// The metadata does not propagate over RPC requests due to issue [#3779].
+///
+/// [#3779]: https://github.com/apache/arrow-rs/issues/3779
```
It isn't at the moment, just prevents the CLI from receiving the metadata. I'll prep a PR in Arrow if you like, as I have a fix working locally.

alamb replied:
> I'll prep a PR in Arrow if you like, as I have a fix working locally.

That would be awesome -- We'll make sure to get it reviewed in short order

stuartcarnie commented on iox_query/src/frontend/influxql.rs:
```diff
@@ -43,6 +51,61 @@ impl SchemaProvider for ContextSchemaProvider {
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        unimplemented!()
```
Will do 👍🏻

My rational was we'd never register this operator as a rule, so it was reasonable to leave these as panicking with `unimplemented` messages, given its sole purpose was as a wrapper around the execution plan.

alamb replied:
I think that is fine too

alamb reviewed (approved):
Looks great @stuartcarnie

wolffcm reviewed (approved):
Still looks good!
